[SPARK-35312][SS] Introduce new Option in Kafka source to specify minimum number of records to read per trigger

satishgopalani opened the pull request:
### What changes were proposed in this pull request?
This patch introduces a new option to specify the minimum number of offsets to read per trigger i.e. minOffsetsPerTrigger and maxTriggerDelay to avoid the infinite wait for the trigger.

This new option will allow skipping trigger/batch when the number of records available in Kafka is low. This is a very useful feature in cases where we have a sudden burst of data at certain intervals in a day and data volume is low for the rest of the day. 
'maxTriggerDelay' option will help to avoid cases of infinite delay in scheduling trigger and the trigger will happen irrespective of records available if the maxTriggerDelay time exceeds the last trigger. It would be an optional parameter with a default value of 15 mins. This option will be only applicable if minOffsetsPerTrigger is set.

minOffsetsPerTrigger option would be optional of course, but once specified it would take precedence over maxOffestsPerTrigger which will be honored only after minOffsetsPerTrigger is satisfied.

### Why are the changes needed?
There are many scenarios where there is a sudden burst of data at certain intervals in a day and data volume is low for the rest of the day. Tunning such jobs is difficult as decreasing trigger processing time increasing the number of batches and hence cluster resource usage and adds to small file issues. Increasing trigger processing time adds consumer lag. This patch tries to address this issue.

### How was this patch tested?
This patch was tested by adding test cases as well as manually on a cluster where the job was running for a full one day with a data burst happening once a day.
Here is the picture of databurst and hence consumer lag:
<img width="1198" alt="Screenshot 2021-04-29 at 11 39 35 PM" src="https://user-images.githubusercontent.com/1044003/116997587-9b2ab180-acfa-11eb-91fd-524802ce3316.png">

This is how the job behaved at burst time running every 4.5 mins (which is the specified trigger time): 
<img width="1154" alt="Burst Time" src="https://user-images.githubusercontent.com/1044003/116997919-12f8dc00-acfb-11eb-9b0a-98387fc67560.png">

This is job behavior during the non-burst time where it is skipping 2 to 3 triggers and running once every 9 to 13.5 mins
<img width="1154" alt="Non Burst Time" src="https://user-images.githubusercontent.com/1044003/116998244-8b5f9d00-acfb-11eb-8340-33d47149ef81.png">

Here are some more stats from the two-run i.e. one normal run and the other with minOffsetsperTrigger set:

| Run | Data Size | Number of Batch Runs | Number of Files |
| ------------- | ------------- |------------- |------------- |
| Normal Run | 54.2 GB | 320 | 21968 |
| Run with minOffsetsperTrigger | 54.2 GB | 120 | 12104 |

pm-satish-gopalani committed:
SPARK-35312: Adding ReadMinRows and CompositeReadLimit to control stream based on minimum rows

satishgopalani commented:
@HeartSaVioR , @bozhang2820, @brkyvz,

I have this new pull with two new ReadLimit options ReadMinRows and CompositeReadLimit as discussed on https://github.com/apache/spark/pull/32434

Could you please review the changes?

cc: @tdas, @zsxwing, @viirya, @gaborgsomogyi, @xuanyuanking

AmplabJenkins commented:
Can one of the admins verify this patch?

HeartSaVioR commented:
ok to test

SparkQA commented:
**[Test build #138905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138905/testReport)** for PR 32653 at commit [`fb14f75`](https://github.com/apache/spark/commit/fb14f75586300b2602346be8854caad80924862e).

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:
```diff
@@ -64,6 +66,15 @@ private[kafka010] class KafkaMicroBatchStream(
+  private[kafka010] val minOffsetPerTrigger = Option(options.get(
+    KafkaSourceProvider.MIN_OFFSET_PER_TRIGGER)).map(_.toLong)
+
+  private[kafka010] val maxTriggerDelayMs =
```
I'd like to see this as a part of `ReadMinRows`, so that other data source implementations leveraging this should also respect the maximum wait time. But let's hear others' voices as well.

satishgopalani replied:
Make sense. But will wait for other reviewers before making changes

satishgopalani replied:
Done

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:
```diff
@@ -95,15 +112,51 @@ private[kafka010] class KafkaMicroBatchStream(
   }
 
+  /** Checks if we need to skip this trigger based on minOffsetsPerTrigger & maxTriggerDelay */
+  private def delayBatch(minLimit: Long,
```
Indentation looks to be off: please refer the style guide on indentation of the method parameters.
https://github.com/databricks/scala-style-guide

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:
```diff
@@ -87,9 +88,17 @@ private[kafka010] class KafkaSource(
+  private[kafka010] val minOffsetPerTrigger =
+    sourceOptions.get(MIN_OFFSET_PER_TRIGGER).map(_.toLong)
+
+  private[kafka010] val maxTriggerDelayMs =
```
Same here.

HeartSaVioR commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/streaming/CompositeReadLimit.java:
```diff
@@ -0,0 +1,61 @@
+ * @since 3.1.2
+ */
+@Evolving
+public final class CompositeReadLimit implements ReadLimit {
```
I'd like to see this be general container of ReadLimits containing `ReadLimit[]` or `List<ReadLimit>`, instead of sticking with minRows and maxRows. e.g. Once we want to do the same for ReadMinFiles we will have to have another class.

HeartSaVioR replied:
Looks to be addressed. Thanks :)

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala:
```diff
@@ -524,6 +528,9 @@ private[kafka010] object KafkaSourceProvider extends Logging {
   private val FAIL_ON_DATA_LOSS_OPTION_KEY = "failondataloss"
   private[kafka010] val MIN_PARTITIONS_OPTION_KEY = "minpartitions"
   private[kafka010] val MAX_OFFSET_PER_TRIGGER = "maxoffsetspertrigger"
+  private[kafka010] val MIN_OFFSET_PER_TRIGGER = "minOffsetsPerTrigger"
```
We use all lowercases here to represent the fact the option name is case insensitive.

HeartSaVioR reviewed (commented):
Probably I need to also discuss with other reviewers on the direction;

1. Include max trigger delay into `ReadMinRows`

I think the functionality is likely essential for every data sources implementing min rows. I don't think we want to leave data source to delay on processing forever.

2. Define `CompositeReadLimit` to contain general ReadLimits

This is to ensure flexibility of the new class `CompositeReadLimit`. If we get more ideas to apply on top of existing read limits, we will end up creating more and more classes. It sounds more sense to me to just have a container class of ReadLimits and let data source handles elements in the class.

@brkyvz @bozhang2820 WDYT?

satishgopalani commented:
@HeartSaVioR Thanks for the review.  It all makes sense to me.
But I will wait for other reviewers to comment and then update PR

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43427/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43427/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43427/

SparkQA commented:
**[Test build #138905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138905/testReport)** for PR 32653 at commit [`fb14f75`](https://github.com/apache/spark/commit/fb14f75586300b2602346be8854caad80924862e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class CompositeReadLimit implements ReadLimit `
  * `public final class ReadMinRows implements ReadLimit `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138905/

pm-satish-gopalani committed:
SPARK-35312: Making CompositeReadLimit generic to accept array of ReadLimit. Moving maxTriggerDelay inside ReadMinRows. Added unit test cases

satishgopalani commented:
@HeartSaVioR 
I have addressed your review comments:
Summary of changes:

- Made CompositeReadLimit generic to accept an array of ReadLimit. 
- Moved maxTriggerDelay inside ReadMinRows
- Added unit test cases to cover for minOffsetsPerTrigger and CompositeReadLimit
- Fixed Scala Style issues and formatting changes

Could you please review the updated pull request?

SparkQA commented:
**[Test build #139028 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139028/testReport)** for PR 32653 at commit [`07e034c`](https://github.com/apache/spark/commit/07e034c61ea8e5011f18d6ce5c34fd5f1905cb4e).

SparkQA commented:
**[Test build #139028 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139028/testReport)** for PR 32653 at commit [`07e034c`](https://github.com/apache/spark/commit/07e034c61ea8e5011f18d6ce5c34fd5f1905cb4e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139028/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43545/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43545/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43545/

HeartSaVioR commented:
Let's ping once again and wait for a couple of days more. I'd like to make sure I'm not the only one thinking of such direction. Thanks for your patience.

@brkyvz @bozhang2820 Kindly reminder.

pm-satish-gopalani committed:
Minor Scala linter fix

SparkQA commented:
**[Test build #139048 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139048/testReport)** for PR 32653 at commit [`b59e02e`](https://github.com/apache/spark/commit/b59e02e131f82f80489c527f202064d3de5f4fb9).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43568/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43568/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43568/

SparkQA commented:
**[Test build #139048 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139048/testReport)** for PR 32653 at commit [`b59e02e`](https://github.com/apache/spark/commit/b59e02e131f82f80489c527f202064d3de5f4fb9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139048/

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:
```diff
@@ -95,15 +114,62 @@ private[kafka010] class KafkaMicroBatchStream(
   override def latestOffset(start: Offset, readLimit: ReadLimit): Offset = {
     val startPartitionOffsets = start.asInstanceOf[KafkaSourceOffset].partitionToOffsets
     latestPartitionOffsets = kafkaOffsetReader.fetchLatestOffsets(Some(startPartitionOffsets))
+    var skipBatch = false
```
Now I see duplicated codes around due to branches handling each type, including CompositeReadLimit which handles both lower and upper hence having same code.

How about changing like below:

```
    val limits: Seq[ReadLimit] = readLimit match {
      case rows: CompositeReadLimit => rows.getReadLimits
      case rows => Seq(rows)
    }

    val offsets = if (limits.exists(_.isInstanceOf[ReadAllAvailable])) {
      // ReadAllAvailable has the highest priority
      latestPartitionOffsets
    } else {
      val lowerLimit = limits.find(_.isInstanceOf[ReadMinRows]).map(_.asInstanceOf[ReadMinRows])
      val upperLimit = limits.find(_.isInstanceOf[ReadMaxRows]).map(_.asInstanceOf[ReadMaxRows])

      lowerLimit.flatMap { limit =>
        // checking if we need to skip batch based on minOffsetPerTrigger criteria
        val skipBatch = delayBatch(
          limit.minRows, latestPartitionOffsets, startPartitionOffsets, limit.maxTriggerDelayMs)
        if (skipBatch) {
          logDebug(
            s"Delaying batch as number of records available is less than minOffsetsPerTrigger")
          Some(startPartitionOffsets)
        } else {
          None
        }
      }.orElse {
        // checking if we need to adjust a range of offsets based on maxOffsetPerTrigger criteria
        upperLimit.map { limit =>
          rateLimit(limit.maxRows(), startPartitionOffsets, latestPartitionOffsets)
        }
      }.getOrElse(latestPartitionOffsets)
    }

    endPartitionOffsets = KafkaSourceOffset(offsets)
    endPartitionOffsets
```

this would require less change when we want to add more read limits in the future.

HeartSaVioR replied:
Btw I guess we can extract the logic of calculating offsets (including delayBatch, of course) based on ReadLimit across DSv1 and DSv2. These codes are mostly duplicated.

Maybe having a new object `KafkaSourceOffsetHelper` and extracting the logic into method(s) there? The content of `getDefaultReadLimit` is another good thing to extract into  as well.

satishgopalani replied:
@HeartSaVioR I had the same thought but I saw rateLimit was copied so kept it same. Was waiting for feedback from community. I will make these changes and also move rateLimit there

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:
```diff
@@ -139,26 +156,78 @@ private[kafka010] class KafkaSource(
-      currentPartitionOffsets.orElse(Some(initialPartitionOffsets)))
+    val currentOffsets = currentPartitionOffsets.orElse(Some(initialPartitionOffsets))
+    val latest = kafkaReader.fetchLatestOffsets(currentOffsets)
+    var skipBatch = false
```
Same here as well.

HeartSaVioR replied:
Something like:

```
  override def latestOffset(startOffset: streaming.Offset, limit: ReadLimit): streaming.Offset = {
    // Make sure initialPartitionOffsets is initialized
    initialPartitionOffsets
    val currentOffsets = currentPartitionOffsets.orElse(Some(initialPartitionOffsets))
    val latest = kafkaReader.fetchLatestOffsets(currentOffsets)
    latestPartitionOffsets = Some(latest)

    val limits: Seq[ReadLimit] = limit match {
      case rows: CompositeReadLimit => rows.getReadLimits
      case rows => Seq(rows)
    }

    val offsets = if (limits.exists(_.isInstanceOf[ReadAllAvailable])) {
      // ReadAllAvailable has the highest priority
      latest
    } else {
      val lowerLimit = limits.find(_.isInstanceOf[ReadMinRows]).map(_.asInstanceOf[ReadMinRows])
      val upperLimit = limits.find(_.isInstanceOf[ReadMaxRows]).map(_.asInstanceOf[ReadMaxRows])

      lowerLimit.flatMap { limit =>
        // checking if we need to skip batch based on minOffsetPerTrigger criteria
        val skipBatch = delayBatch(
          limit.minRows, latest, currentOffsets.get, limit.maxTriggerDelayMs)
        if (skipBatch) {
          logDebug(
            s"Delaying batch as number of records available is less than minOffsetsPerTrigger")
          // Pass same current offsets as output to skip trigger
          Some(currentOffsets.get)
        } else {
          None
        }
      }.orElse {
        // checking if we need to adjust a range of offsets based on maxOffsetPerTrigger criteria
        upperLimit.map { limit =>
          rateLimit(limit.maxRows, currentPartitionOffsets.getOrElse(initialPartitionOffsets),
            latest)
        }
      }.getOrElse(latest)
    }

    currentPartitionOffsets = Some(offsets)
    logDebug(s"GetOffset: ${offsets.toSeq.map(_.toString).sorted}")
    KafkaSourceOffset(offsets)
  }
```

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -321,6 +321,231 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+      .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+      .option("kafka.metadata.max.age.ms", "1")
+      .option("minOffsetsPerTrigger", 15)
+      .option("maxTriggerDelay", "10s")
```
Let's pick smaller value like 5 seconds or similar. It won't introduce flakiness and less time duration to run.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -321,6 +321,231 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+      // and maxTriggerDelay is not expired
+      AdvanceManualClock(100),
+      waitUntilBatchProcessed,
+      CheckAnswer(1, 10, 100, 101, 102, 103, 104, 105, 106, 107,
```
Please use CheckNewAnswer here and below to assert only new output if that works for the test.

If CheckNewAnswer works without waitUntilBatchProcessed, even better.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -321,6 +321,231 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+      AdvanceManualClock(100),
+      waitUntilBatchProcessed,
+      // Running batch now as number of records is greater than minOffsetsPerTrigger
+      // 2 from smallest, 10 more from middle, 20 more from biggest
```
Once CheckNewAnswer works we don't need the code comment for expected values.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -321,6 +321,231 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+
+    val clock = new StreamManualClock
+
+    val waitUntilBatchProcessed = AssertOnQuery { q =>
```
Let's extract this to `private def` and reuse across tests in this file.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -321,6 +321,231 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+    }
+  }
+
+  test("compositeReadLimit") {
```
Same comments here.

satishgopalani replied:
Updated

satishgopalani commented:
@HeartSaVioR Will work on review comments and reshuffle to make even older functions generic and avoid code duplication

HeartSaVioR commented:
You can skip the deduplication between DSv1 and DSv2. That's something we are behind for a long time and we can deal with it with further PRs deduplicating more things.

HeartSaVioR commented:
I'm sorry to give a wrong signal before, but let's not refactor on existing methods. That would make us harder to review. Let's focus on your changes.

pm-satish-gopalani committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-35312

pm-satish-gopalani committed:
Implementing review comments

satishgopalani commented:
@HeartSaVioR 
Thanks for your detailed review.
I have updated the pull request implementing all the review comments.
I have not implemented any code deduplication as per your last comment and I have kept a copy of delayBatch in both classes. I think deduplication can be tracked as a separate PR.

SparkQA commented:
**[Test build #139292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139292/testReport)** for PR 32653 at commit [`046d751`](https://github.com/apache/spark/commit/046d751c005376aad290d96d80478c64da8c4c29).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43816/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43816/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43816/

SparkQA commented:
**[Test build #139292 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139292/testReport)** for PR 32653 at commit [`046d751`](https://github.com/apache/spark/commit/046d751c005376aad290d96d80478c64da8c4c29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139292/

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:
```diff
@@ -95,15 +114,65 @@ private[kafka010] class KafkaMicroBatchStream(
+      latestOffsets: Map[TopicPartition, Long],
+      currentOffsets: Map[TopicPartition, Long],
+      maxTriggerDelayMs: Long): Boolean = {
+    // Checking first if the maxbatchDelay time has passed
```
nit: It won't hurt if we only call `System.currentTimeMillis()` once and reuse it.

HeartSaVioR reviewed (approved):
+1 I added the nit comment but I think it's super minor.

I'll retrigger the test and merge once test passes.

If Jenkins doesn't like me, I'll need the help from @satishgopalani to retrigger the Github Action test. If then we can address the nit and push the commit to trigger the test.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #139460 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139460/testReport)** for PR 32653 at commit [`046d751`](https://github.com/apache/spark/commit/046d751c005376aad290d96d80478c64da8c4c29).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43983/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43983/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43983/

SparkQA commented:
**[Test build #139460 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139460/testReport)** for PR 32653 at commit [`046d751`](https://github.com/apache/spark/commit/046d751c005376aad290d96d80478c64da8c4c29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139460/

HeartSaVioR commented:
Thanks! Merging to master.

HeartSaVioR commented:
Thanks @satishgopalani for your first contribution to Apache Spark community! I merged this to master.

satishgopalani commented:
Thanks a lot, @HeartSaVioR for your review and inputs. It helped me a lot. 
Looking forward to many more contributions to the Apache Spark community!
